components/profile-controller/pkg/apis/kubeflow/v1beta1/profile_types.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2019 The Kubeflow Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Plugin is for customize actions on different platform.
+type Plugin struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              *runtime.RawExtension `json:"spec,omitempty"`
+}
+
+// ProfileSpec defines the desired state of Profile
+type ProfileSpec struct {
+	// The profile owner
+	Owner   rbacv1.Subject `json:"owner,omitempty"`
+	Plugins []Plugin       `json:"plugins,omitempty"`
+	// Resourcequota that will be applied to target namespace
+	ResourceQuotaSpec v1.ResourceQuotaSpec `json:"resourcequotaspec,omitempty"`
+}
+
+type ProfileState string
```

Comment:
I would suggest we avoid using enum fields in status, as they break if we add new possible values.
Instead, the K8s api convention advises one to use Conditions.
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md

Reply:
ProfileState is saved as plain string, adding new values should not change existing messages.
We can use conditions in followup PRs if needed.

Reply:
@kunmingg I'm a bit hesitant to go forward with a Phase enum-type status, because this is the pattern that Kubernetes used and then decided explicitly against it, in favor of conditions.
See [this section](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) for more information: 

> pattern of using phase is deprecated. Newer API types should use conditions instead. Phase was essentially a state-machine enumeration field, that contradicted system-design principles and hampered evolution, since adding new enum values breaks backward compatibility. 


/cc @jlewi @kkasravi

Reply:
+1 to using conditions. 

@kunmingg are there problems with using conditions rather than enums?